### PR TITLE
udev: Fix for devices implementing BTN_DPAD_* codes being classified as ID_INPUT_KEY devices.

### PR DESCRIPTION
**Pull request by nrlewis117**
Fixes: #5703 

At present, devices implementing the BTN_DPAD_UP/DOWN/LEFT/RIGHT
codes will be incorrectly classified as key devices.  This causes
devices respecting the Linux gamepad spec (such as the DS3 as of
kernel 4.12) to be classified as keyboards by X11.

This is caused by the test_key function checking all codes on
[KEY_OK, BTN_TRIGGER_HAPPY).  Unfortunately the BTN_DPAD_* codes
are placed between KEY_LIGHTS_TOGGLE and KEY_ALS_TOGGLE.  This
patch splits the upper key block check into the block before and
after the BTN_DPAD_* codes.  An array is used to avoid dedicated loops
per block in the event that more event codes are added in
the future.

### CODE REVIEW

**Comment by nrlewis117** [comment]
I was under the assumption that BTN_DPAD_UP and BTN_ALS_TOGGLE were safe, considering those were added in kernels 3.11 and 3.13 respectively, and the minimum supported kernel by systemd is 3.13.

File: src/udev/udev-builtin-input_id.c [hunk 1]
```diff
@@ -44,6 +44,12 @@
 #define LONG(x) ((x)/BITS_PER_LONG)
 #define test_bit(bit, array)    ((array[LONG(bit)] >> OFF(bit)) & 1)
 
+/* key code ranges above BTN_MISC (start is inclusive, stop is exclusive)*/
+static const unsigned high_key_blocks[] = {
+        KEY_OK, BTN_DPAD_UP,
+        KEY_ALS_TOGGLE, BTN_TRIGGER_HAPPY
+};
```
Comment (martinpitt) [inline src/udev/udev-builtin-input_id.c]: This and the loop below would be more elegant/simpler as a list of struct. But just thinking aloud, it's good enough as-is IMHO.

File: src/udev/udev-builtin-input_id.c [hunk 2]
```diff
@@ -244,6 +250,7 @@ static bool test_key(struct udev_device *dev,
                      const unsigned long* bitmask_key,
                      bool test) {
         unsigned i;
+        unsigned j;
```
Comment (martinpitt) [inline src/udev/udev-builtin-input_id.c]: Please don't declare this here, move it into the inner `for i..` loop as `for (unsigned j = 0; ...)`. Let's not make the scope of these single-letter variables larger than absolutely necessary.

**Comment by nrlewis117** [comment]
I've updated the commit with with your change requests.

**Comment by Teknoman117** [comment]
Seems the tests have all completed successfully now.

**Comment by nrlewis117** [comment]
Looks like all of the tests have passed.

**Pull request merged**